gneiss/balances.py

```diff
@@ -90,51 +102,63 @@ def balance_basis(tree_node):
 
 
 def _count_matrix(treenode):
-    n_tips = 0
-    nodes = list(treenode.levelorder(include_self=True))
-    # fill in the Ordered dictionary. Note that the
-    # elements of this Ordered dictionary are
-    # dictionaries.
-    counts = OrderedDict()
-    columns = ['k', 'r', 'l', 't', 'tips']
-    for n in nodes:
-        if n not in counts:
-            counts[n] = {}
-        for c in columns:
-            counts[n][c] = 0
-
-    # fill in r and l.  This is done in reverse level order.
-    for n in nodes[::-1]:
+    node_count = 0
+    for n in treenode.postorder(include_self=True):
+        node_count += 1
         if n.is_tip():
-            counts[n]['tips'] = 1
-            n_tips += 1
-        elif len(n.children) == 2:
-            lchild = n.children[0]
-            rchild = n.children[1]
-            counts[n]['r'] = counts[rchild]['tips']
-            counts[n]['l'] = counts[lchild]['tips']
-            counts[n]['tips'] = counts[n]['r'] + counts[n]['l']
+            n._tip_count = 1
         else:
-            raise ValueError("Not a strictly bifurcating tree!")
-
-    # fill in k and t
-    for n in nodes:
-        if n.parent is None:
-            counts[n]['k'] = 0
-            counts[n]['t'] = 0
-            continue
-        elif n.is_tip():
+            try:
+                left, right = n.children
+            except:
+                raise ValueError("Not a strictly bifurcating tree!")
+            n._tip_count = left._tip_count + right._tip_count
+
+    # TODO: use recarray
+    # col 0 -> right counts
+    # col 1 -> left counts
+    # col 2 -> k
```

Comment:
Ah, okay, thanks. So in the tree as a whole, this makes much more sense.
Will kick this around more tomorrow
On Jul 11, 2016 12:04 AM, "Jamie Morton" <notifications@github.com> wrote:

> In gneiss/balances.py
> <https://github.com/biocore/gneiss/pull/8#discussion_r70209614>:
>
> > -    for n in nodes:
> > -        if n.parent is None:
> > -            counts[n]['k'] = 0
> > -            counts[n]['t'] = 0
> > -            continue
> > -        elif n.is_tip():
> > +            try:
> > +                left, right = n.children
> > +            except:
> > +                raise ValueError("Not a strictly bifurcating tree!")
> > +            n._tip_count = left._tip_count + right._tip_count
> > +
> > +    # TODO: use recarray
> > +    # col 0 -> right counts
> > +    # col 1 -> left counts
> > +    # col 2 -> k
>
> k is more concerned about the tips to the right of 4 and t is more
> concerned about the tips left of 4. I realize that the orientation is a
> little confusing.
>
> The reason why we want to do this, is because at each internal node of the
> tree, all of the tips of the tree fall into 2 categories.
> Tips to the right of the node, but are not contained in the subtree (k)
> Tips that are descendants of the node, but are in the right subtree (r)
> Tips that are descendants of the node, but are in the left subtree (l)
> Tips to the left of the node, but are not contained in the subtree (t)
>
> So these 4 numbers should always add up to
> k+r+l+t = N where N is the total number of tips in the tree.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/gneiss/pull/8/files/23ad7ae6d93099e1a6575efac2e7500ac10833dd#r70209614>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAc8smfxtJlVi6EYW9jryVBT9jePhT5eks5qUesEgaJpZM4JI_gR>
> .
>